mesonbuild/mesonlib/universal.py

```diff
@@ -1887,34 +1886,57 @@ def __call__(self, parser: argparse.ArgumentParser, namespace: argparse.Namespac
 
 
 def get_wine_shortpath(winecmd: T.List[str], wine_paths: T.Sequence[str]) -> str:
-    """Get A short version of @wine_paths to avoid reaching WINEPATH number
-    of char limit.
-    """
+    '''
+    WINEPATH size is limited to 1024 bytes which can easily be exceeded when
+    adding the path to every dll inside build directory. See
+    https://bugs.winehq.org/show_bug.cgi?id=45810.
+
+    To shorthen it as much as possible we convert paths to Windows shortpath
+    (e.g. "/usr/x86_64-w64-mingw32/lib" to "Z:\\usr\\X86_~FWL\\lib").
 
+    This limitation reportedly have been fixed with wine >= 6.4
+    '''
+
+    # Remove duplicates
     wine_paths = list(OrderedSet(wine_paths))
 
-    getShortPathScript = '%s.bat' % str(uuid.uuid4()).lower()[:5]
-    with open(getShortPathScript, mode='w', encoding='utf-8') as f:
-        f.write("@ECHO OFF\nfor %%x in (%*) do (\n echo|set /p=;%~sx\n)\n")
-        f.flush()
+    # Check if it's already short enough
+    wine_path = ';'.join(wine_paths)
+    if len(wine_path) <= 1024:
+        return wine_path
+
+    # Check if we have wine >= 6.4
     try:
-        with open(os.devnull, 'w', encoding='utf-8') as stderr:
-            wine_path = subprocess.check_output(
-                winecmd +
-                ['cmd', '/C', getShortPathScript] + wine_paths,
-                stderr=stderr).decode('utf-8')
+        stdout = subprocess.check_output([winecmd, '--version'], encoding='utf-8')
+        match = re.search(r'([0-9][0-9\.]+)', stdout)
```

Comment:
`programs.ExternalProgram('wine', winecmd).get_version()`  adds a much nicer UX for checking versions, including a single unified regex.

Reply:
But it takes an interpreter in args which we don't have. Would have to refactor that further.

Reply:
Yeah. :(

Reply:
Made interpreter optional and use Popen_safe() in the case it's not provided.

Reply:
Can we split that into its own commit? :)

Reply:
done.